### PR TITLE
Object input helper functions - Closes #243

### LINKED ISSUE #243: Add options object to API functions
The helper API functions should take 1 maximum 2 parameters as helper. Every other parameter that is accepted by core should be able to be passed in as object

Part of #139

### PR DESCRIPTION
**Pull request by Tosch110**
Closes #243

### CODE REVIEW

File: src/api/liskApi.js [hunk 1]
```diff
@@ -243,153 +258,190 @@ LiskAPI.prototype.getAddressFromSecret = function getAddressFromSecret(secret) {
...
  */
 
-LiskAPI.prototype.listActiveDelegates = function listActiveDelegates(limit, callback) {
-	this.sendRequest(GET, 'delegates/', { limit }, result => callback(result));
+LiskAPI.prototype.listActiveDelegates = function listActiveDelegates(
+	limit, optionsOrCallback, callbackIfOptions,
+) {
+	const callback = callbackIfOptions || optionsOrCallback;
+	const data = this.constructRequestData({ limit }, optionsOrCallback);
+	return this.sendRequest(GET, 'delegates/', data, callback);
```
Comment (willclarktech) [inline src/api/liskApi.js]: We should be consistent about whether we add a trailing slash or not.

File: src/api/liskApi.js [hunk 2]
```diff
@@ -243,153 +258,190 @@ LiskAPI.prototype.getAddressFromSecret = function getAddressFromSecret(secret) {
...
 
-LiskAPI.prototype.listStandbyDelegates = function listStandbyDelegates(limit, callback) {
-	const standByOffset = 101;
-
-	this.sendRequest(GET, 'delegates/', { limit, orderBy: 'rate:asc', offset: standByOffset }, result => callback(result));
+LiskAPI.prototype.listStandbyDelegates = function listStandbyDelegates(
+	limit, optionsOrCallback, callbackIfOptions,
+) {
+	const offset = typeof optionsOrCallback !== 'function' && optionsOrCallback.offset ? optionsOrCallback.offset : 101;
+	const orderBy = typeof optionsOrCallback !== 'function' && optionsOrCallback.orderBy ? optionsOrCallback.orderBy : 'rate:asc';
```
Comment (willclarktech) [inline src/api/liskApi.js]:
```js
const { offset=101, orderBy='rate:asc' } = optionsOrCallback;
```

File: src/api/liskApi.js [hunk 3]
```diff
@@ -243,153 +258,190 @@ LiskAPI.prototype.getAddressFromSecret = function getAddressFromSecret(secret) {
...
-LiskAPI.prototype.generateAccount = function generateAccount(secret, callback) {
-	const keys = LiskJS.crypto.getPrivateAndPublicKeyFromSecret(secret);
-	callback(keys);
-	return this;
+LiskAPI.prototype.getAccount = function getAccount(
+	address, optionsOrCallback, callbackIfOptions,
+) {
+	const callback = callbackIfOptions || optionsOrCallback;
+	const data = this.constructRequestData({ address }, optionsOrCallback);
+	return this.sendRequest(GET, 'accounts', data, callback);
```
Comment (willclarktech) [inline src/api/liskApi.js]: We can get rid of more boilerplate:
```js
LiskAPI.prototype.wrapSendRequest = function wrapSendRequest (
  method, endpoint, specialData, optionsOrCallback, callbackIfOptions,
) {
  const callback = callbackIfOptions || optionsOrCallback;
  const data = this.constructRequestData(specialData, optionsOrCallback);
  return this.sendRequest(method, endpoint, data, callback);
};

Lisk.API.prototype.getAccount = function getAccount(
  address, optionsOrCallback, callbackIfOptions,
) {
  return this.wrapSendRequest(GET, 'accounts', { address }, optionsOrCallback, callbackIfOptions);
};
```

File: src/api/liskApi.js [hunk 4]
```diff
@@ -417,21 +471,24 @@ LiskAPI.prototype.sendLSK = function sendLSK(recipient, amount, secret, secondSe
...
  */
 
 LiskAPI.prototype.getMultisignatureTransaction = function getMultisignatureTransaction(
-	transactionId, callback,
+	id, optionsOrCallback, callbackIfOptions,
 ) {
-	this.sendRequest(GET, 'transactions/multisignatures/get', { id: transactionId }, result => callback(result));
+	const callback = callbackIfOptions || optionsOrCallback;
+	const data = this.constructRequestData({ id }, optionsOrCallback);
+	this.sendRequest(GET, 'transactions/multisignatures/get', data, callback);
```
Comment (willclarktech) [inline src/api/liskApi.js]: `return`

File: test/api/liskApi.js [hunk 5]
```diff
@@ -400,7 +400,7 @@ describe('Lisk.api()', () => {
 
 			LSK.searchDelegateByUsername('oliver', callback);
 
-			(LSK.sendRequest.calledWith('GET', 'delegates/search/', options)).should.be.true();
+			(LSK.sendRequest.calledWith(GET, 'delegates/search/', options)).should.be.true();
```
Comment (willclarktech) [inline test/api/liskApi.js]: Prefer `calledWithExactly`

File: test/api/liskApi.js [hunk 6]
```diff
@@ -538,16 +538,17 @@ describe('Lisk.api()', () => {
 		it('should list transactions of a defined account', () => {
 			const callback = sinon.spy();
 			const address = '12731041415715717263L';
+			const optionAddress = '15731041415715717263L';
```
Comment (willclarktech) [inline test/api/liskApi.js]: Prefer naming these `senderAddress` and `recipientAddress`.

File: test/api/liskApi.js [hunk 7]
```diff
@@ -981,4 +965,30 @@ describe('Lisk.api()', () => {
...
+
+		it('should construct with variable and callback', () => {
+			const address = '123';
+			const expectedObject = {
+				address: '123',
+			};
+			const createObject = LSK.constructRequestData({ address }, () => { return '123'; });
+			(createObject).should.be.eql(expectedObject);
+		});
+	});
```
Comment (willclarktech) [inline test/api/liskApi.js]: We need a test showing that options can be overridden.

File: test/api/liskApi.js [hunk 8]
```diff
@@ -981,4 +965,30 @@ describe('Lisk.api()', () => {
 			(requestObject).should.be.eql(expectedObject);
 		});
 	});
+
+	describe('#constructRequestData @now', () => {
```
Comment (willclarktech) [inline test/api/liskApi.js]: `@now`

File: src/api/liskApi.js (hunk 3 above)
Comment (willclarktech) [inline src/api/liskApi.js]: Or a higher order function solution:
```js
LiskAPI.prototype.wrapSendRequest = function wrapSendRequest(method, endpoint, getDataFn) {
  return (value, optionsOrCallback, callbackIfOptions) => {
    const callback = callbackIfOptions || optionsOrCallback;
    const data = this.constructRequestData(getDataFn(value, optionsOrCallback), optionsOrCallback));
    return this.sendRequest(method, endpoint, data, callback);
  };
};
LiskAPI.protototype.getAccount = this.wrapSendRequest(GET, 'accounts', address => ({ address }));
```

File: src/api/liskApi.js [hunk 9]
```diff
@@ -258,191 +243,120 @@ LiskAPI.prototype.getAddressFromSecret = function getAddressFromSecret(secret) {
 /**
  * @method getAccount
  * @param address
- * @param optionsOrCallback
- * @param callbackIfOptions
```
Comment (willclarktech) [inline src/api/liskApi.js]: The wrapped function takes these arguments:
```js
function wrappedSendRequest(value, optionsOrCallback, callbackIfOptions)
```

File: src/api/liskApi.js [hunk 10]
```diff
@@ -258,191 +243,120 @@ LiskAPI.prototype.getAddressFromSecret = function getAddressFromSecret(secret) {
...
  */
 
-LiskAPI.prototype.getAccount = function getAccount(
-	address, optionsOrCallback, callbackIfOptions,
-) {
-	const callback = callbackIfOptions || optionsOrCallback;
-	const data = this.constructRequestData({ address }, optionsOrCallback);
-	return this.sendRequest(GET, 'accounts', data, callback);
-};
+LiskAPI.prototype.getAccount = privateApi.wrapSendRequest(GET, 'accounts', address => ({ address }));
```
Comment (willclarktech) [inline src/api/liskApi.js]: 👌

File: src/api/liskApi.js [hunk 11]
```diff
@@ -258,191 +243,120 @@ LiskAPI.prototype.getAddressFromSecret = function getAddressFromSecret(secret) {
...
+
+/**
+ * @method listMultisignatureTransactions
+ * @param data
+ * @param callback
+ *
+ * @return API object
+ */
+
+LiskAPI.prototype.listMultisignatureTransactions = privateApi.wrapSendRequest(GET, 'transactions/multisignatures', data => ({ data }));
```
Comment (willclarktech) [inline src/api/liskApi.js]: Should be `data => data`?

File: src/api/privateApi.js [hunk 12]
```diff
@@ -227,6 +227,37 @@ function sendRequestPromise(requestMethod, requestType, options) {
...
+ * @method constructRequestData
+ * @param providedObject
+ * @param optionsOrCallback
+ *
+ * @return request object
+ */
+
+const constructRequestData = (
+	providedObject, optionsOrCallback,
+) => {
```
Comment (willclarktech) [inline src/api/privateApi.js]: This probably fits on one line now.

File: src/api/privateApi.js [hunk 13]
```diff
@@ -227,6 +227,37 @@ function sendRequestPromise(requestMethod, requestType, options) {
...
+ *
+ * @return request object
+ */
+
+const wrapSendRequest = (method, endpoint, getDataFn) =>
+	function wrappedSendRequest(value, optionsOrCallback, callbackIfOptions) {
+		const callback = callbackIfOptions || optionsOrCallback;
+		const data = constructRequestData(getDataFn(value, optionsOrCallback), optionsOrCallback);
+		return this.sendRequest(method, endpoint, data, callback);
+	};
```
Comment (willclarktech) [inline src/api/privateApi.js]: ❤️

File: test/api/liskApi.js [hunk 14]
```diff
@@ -693,8 +693,8 @@ describe('Lisk.api()', () => {
 			LSK.getAccount(address, callback);
 
 			(LSK.sendRequest.calledWith(GET, 'accounts', options)).should.be.true();
-			(callback.called).should.be.true();
-			(callback.calledWith(expectedResponse)).should.be.true();
+			// (callback.called).should.be.true();
+			// (callback.calledWith(expectedResponse)).should.be.true();
```
Comment (willclarktech) [inline test/api/liskApi.js]: Remove if not needed.

File: src/api/liskApi.js [hunk 15]
```diff
@@ -258,191 +243,120 @@ LiskAPI.prototype.getAddressFromSecret = function getAddressFromSecret(secret) {
...
-LiskAPI.prototype.listStandbyDelegates = function listStandbyDelegates(
-	limit, optionsOrCallback, callbackIfOptions,
-) {
-	const offset = typeof optionsOrCallback !== 'function' && optionsOrCallback.offset ? optionsOrCallback.offset : 101;
-	const orderBy = typeof optionsOrCallback !== 'function' && optionsOrCallback.orderBy ? optionsOrCallback.orderBy : 'rate:asc';
-	const callback = callbackIfOptions || optionsOrCallback;
-	const data = this.constructRequestData({ limit, offset, orderBy }, optionsOrCallback);
-	return this.sendRequest(GET, 'delegates/', data, callback);
-};
+LiskAPI.prototype.listStandbyDelegates = privateApi.wrapSendRequest(GET, 'delegates', limit => ({ limit }));
```
Comment (willclarktech) [inline src/api/liskApi.js]: This needs to incorporate `offset` and `orderBy` (it's the reason `getDataFn` takes two arguments).

File: test/api/liskApi.js [hunk 16]
```diff
@@ -275,9 +275,21 @@ describe('Lisk.api()', () => {
 			const options = { limit: '1', orderBy: 'rate:asc', offset: 101 };
 			sinon.stub(LSK, 'sendRequest').callsArgWith(3, expectedResponse);
 
-			LSK.listStandbyDelegates('1', callback);
+			LSK.listStandbyDelegates('1', options, callback);
+
+			(LSK.sendRequest.calledWith(GET, 'delegates', options)).should.be.true();
+			(callback.called).should.be.true();
+			(callback.calledWith(expectedResponse)).should.be.true();
```
Comment (willclarktech) [inline test/api/liskApi.js]: Prefer `calledWithExactly`. (Also below.)

File: test/api/liskApi.js [hunk 17]
```diff
@@ -275,9 +275,21 @@ describe('Lisk.api()', () => {
...
+		});
+
+		it('should list standby delegates without specifying offset or orderBy', () => {
+			const callback = sinon.spy();
+			const options = {};
+			sinon.stub(LSK, 'sendRequest').callsArgWith(3, expectedResponse);
+
+			LSK.listStandbyDelegates('1', options, callback);
 
-			(LSK.sendRequest.calledWith(GET, 'delegates/', options)).should.be.true();
```
Comment (willclarktech) [inline test/api/liskApi.js]: Since we're stubbing `sendRequest`, we also need to test what it's called with.

File: test/api/liskApi.js [hunk 18]
```diff
@@ -695,4 +691,31 @@ describe('Lisk.api()', () => {
 			});
 		});
 	});
+
+
+	describe('#constructRequestData', () => {
+		it('should construct optional request data for API helper functions', () => {
+			const address = '123';
+			const requestData = {
+				limit: '123',
```
Comment (willclarktech) [inline test/api/liskApi.js]: Let's use different numbers for different pieces of data.

File: test/api/liskApi.js [hunk 19]
```diff
@@ -695,4 +691,31 @@ describe('Lisk.api()', () => {
...
+				offset: 5,
+			};
+			const expectedObject = {
+				address: '123',
+				limit: '123',
+				offset: 5,
+			};
+			const createObject = privateApi.constructRequestData({ address }, requestData);
```
Comment (willclarktech) [inline test/api/liskApi.js]: We need a test where options and data conflict.

File: test/api/liskApi.js [hunk 20]
```diff
@@ -695,4 +691,31 @@ describe('Lisk.api()', () => {
...
+			(createObject).should.be.eql(expectedObject);
+		});
+
+		it('should construct with variable and callback', () => {
+			const address = '123';
+			const expectedObject = {
+				address: '123',
+			};
+			const createObject = privateApi.constructRequestData({ address }, () => { return '123'; });
```
Comment (willclarktech) [inline test/api/liskApi.js]: Prefer `() => '123'` arrow body shorthand when there's just a single return statement. But using '123' here is confusing since we already have it for `address`.

File: test/api/liskApi.js [hunk 21]
```diff
@@ -695,4 +691,31 @@ describe('Lisk.api()', () => {
...
+		it('should construct optional request data for API helper functions', () => {
+			const address = '123';
+			const requestData = {
+				limit: '123',
+				offset: 5,
+			};
+			const expectedObject = {
+				address: '123',
+				limit: '123',
+				offset: 5,
```
Comment (willclarktech) [inline test/api/liskApi.js]: Reuse variables instead of magic strings/numbers.

File: src/transactions/utils.js [hunk 22]
```diff
@@ -12,7 +12,7 @@ const prepareTransaction = (transaction, keys, secondSecret) => {
 		signature: crypto.sign(transaction, keys),
 	});
 
-	const signedTransaction = secondSecret && transaction.type !== 1
+	const signedTransaction = (typeof secondSecret === 'string' && transaction.type !== 1)
```
Comment (willclarktech) [inline src/transactions/utils.js]: This addresses #285 and should be in a separate pull request with tests.

File: test/api/liskApi.js [hunk 23]
```diff
@@ -239,14 +241,14 @@ describe('Lisk.api()', () => {
...
 			const callback = sinon.spy();
-			const options = { limit: '1' };
+			const options = { limit: requestDataLimit };
 			sinon.stub(LSK, 'sendRequest').callsArgWith(3, expectedResponse);
 
-			LSK.listActiveDelegates('1', callback);
+			LSK.listActiveDelegates(requestDataLimit, callback);
 
-			(LSK.sendRequest.calledWith(GET, 'delegates/', options)).should.be.true();
+			(LSK.sendRequest.calledWith(GET, 'delegates', options)).should.be.true();
```
Comment (willclarktech) [inline test/api/liskApi.js]: `Exactly`

File: test/api/liskApi.js [hunk 24]
```diff
@@ -272,14 +274,27 @@ describe('Lisk.api()', () => {
 
 		it('should list standby delegates', () => {
 			const callback = sinon.spy();
-			const options = { limit: '1', orderBy: 'rate:asc', offset: 101 };
+			const options = { limit: requestDataLimit, orderBy: 'rate:asc', offset: requestOffset };
```
Comment (willclarktech) [inline test/api/liskApi.js]: We should use `offset` and `orderBy` values which are different from the defaults in this test.

File: test/api/liskApi.js [hunk 25]
```diff
@@ -272,14 +274,27 @@ describe('Lisk.api()', () => {
...
 			sinon.stub(LSK, 'sendRequest').callsArgWith(3, expectedResponse);
 
-			LSK.listStandbyDelegates('1', callback);
+			LSK.listStandbyDelegates(requestDataLimit, options, callback);
 
-			(LSK.sendRequest.calledWith(GET, 'delegates/', options)).should.be.true();
+			(LSK.sendRequest.calledWith(GET, 'delegates', options)).should.be.true();
```
Comment (willclarktech) [inline test/api/liskApi.js]: `Exactly`

File: test/api/liskApi.js [hunk 26]
```diff
@@ -11,6 +11,8 @@ describe('Lisk.api()', () => {
 		publicKey: '5d036a858ce89f844491762eb89e2bfbd50a4a0a0da658e4b2628b25b117ae09',
 		address: '18160565574430594874L',
 	};
+	const requestDataLimit = 10;
+	const requestOffset = 101;
```
Comment (willclarktech) [inline test/api/liskApi.js]: Let's make clear whether these are defaults or not with the name.

File: test/api/liskApi.js [hunk 27]
```diff
@@ -272,14 +274,27 @@ describe('Lisk.api()', () => {
...
 
-			(LSK.sendRequest.calledWith(GET, 'delegates/', options)).should.be.true();
+			(LSK.sendRequest.calledWith(GET, 'delegates', options)).should.be.true();
 			(callback.called).should.be.true();
-			(callback.calledWith(expectedResponse)).should.be.true();
+			(callback.calledWithExactly(expectedResponse)).should.be.true();
+			LSK.sendRequest.restore();
+		});
+
+		it('should list standby delegates without specifying offset or orderBy', () => {
```
Comment (willclarktech) [inline test/api/liskApi.js]: `with a default offset and ordering when not specified`

File: test/api/liskApi.js [hunk 28]
```diff
@@ -272,14 +274,27 @@ describe('Lisk.api()', () => {
...
+			const callback = sinon.spy();
+			const options = {};
+			sinon.stub(LSK, 'sendRequest').callsArgWith(3, expectedResponse);
+
+			LSK.listStandbyDelegates(requestDataLimit, options, callback);
+
+			(LSK.sendRequest.calledWith(GET, 'delegates', { limit: requestDataLimit, orderBy: 'rate:asc', offset: requestOffset })).should.be.true();
```
Comment (willclarktech) [inline test/api/liskApi.js]: `Exactly`

File: test/api/liskApi.js [hunk 29]
```diff
@@ -301,14 +316,14 @@ describe('Lisk.api()', () => {
...
 		it('should find a delegate by name', () => {
 			const callback = sinon.spy();
-			const options = { q: 'oliver' };
+			const options = { username: 'oliver' };
 			sinon.stub(LSK, 'sendRequest').callsArgWith(3, expectedResponse);
 
 			LSK.searchDelegateByUsername('oliver', callback);
 
-			(LSK.sendRequest.calledWith('GET', 'delegates/search/', options)).should.be.true();
+			(LSK.sendRequest.calledWith(GET, 'delegates/search', options)).should.be.true();
```
Comment (willclarktech) [inline test/api/liskApi.js]: `Exactly`

File: test/api/liskApi.js [hunk 30]
```diff
@@ -239,14 +241,14 @@ describe('Lisk.api()', () => {
...
 			sinon.stub(LSK, 'sendRequest').callsArgWith(3, expectedResponse);
 
-			LSK.listActiveDelegates('1', callback);
+			LSK.listActiveDelegates(requestDataLimit, callback);
 
-			(LSK.sendRequest.calledWith(GET, 'delegates/', options)).should.be.true();
+			(LSK.sendRequest.calledWith(GET, 'delegates', options)).should.be.true();
 			(callback.called).should.be.true();
-			(callback.calledWith(expectedResponse)).should.be.true();
+			(callback.calledWithExactly(expectedResponse)).should.be.true();
```
Comment (willclarktech) [inline test/api/liskApi.js]: If you specify the callback in the previous `sendRequest` assertion this will become obsolete (also true for other tests below).

File: test/api/liskApi.js [hunk 31]
```diff
@@ -695,4 +694,45 @@ describe('Lisk.api()', () => {
 			});
 		});
 	});
+
+
+	describe('#constructRequestData @now', () => {
```
Comment (willclarktech) [inline test/api/liskApi.js]: `@now`

File: test/api/liskApi.js [hunk 32]
```diff
@@ -695,4 +694,45 @@ describe('Lisk.api()', () => {
...
+		const address = defaultAddress.address;
+		const requestData = {
+			limit: requestDataLimit,
+			offset: requestOffset,
+		};
+		const expectedObject = {
+			address: '18160565574430594874L',
```
Comment (willclarktech) [inline test/api/liskApi.js]: `address`

File: test/api/liskApi.js [hunk 33]
```diff
@@ -695,4 +694,45 @@ describe('Lisk.api()', () => {
...
+			limit: requestDataLimit,
+			offset: requestOffset,
+		};
+		const conflictObject = {
+			address: '18160565574430594874',
```
Comment (willclarktech) [inline test/api/liskApi.js]: Did you mean to leave the `L` off the end here?

File: test/api/liskApi.js [hunk 34]
```diff
@@ -695,4 +694,45 @@ describe('Lisk.api()', () => {
 			});
 		});
 	});
+
+
+	describe('#constructRequestData @now', () => {
+		const address = defaultAddress.address;
+		const requestData = {
```
Comment (willclarktech) [inline test/api/liskApi.js]: `optionsObject`

File: test/api/liskApi.js [hunk 35]
```diff
@@ -695,4 +694,45 @@ describe('Lisk.api()', () => {
...
+			limit: 0,
+			offset: 5,
+		};
+		const resolvedConflictObject = {
+			address: '18160565574430594874',
+			limit: requestDataLimit,
+			offset: requestOffset,
+		};
+
+		it('should construct optional request data for API helper functions', () => {
```
Comment (willclarktech) [inline test/api/liskApi.js]: `should merge a data object with an options object`

File: test/api/liskApi.js [hunk 36]
```diff
@@ -695,4 +694,45 @@ describe('Lisk.api()', () => {
...
+			const createObject = privateApi.constructRequestData({ address }, requestData);
+			(createObject).should.be.eql(expectedObject);
+		});
+
+		it('should construct with variable and callback', () => {
```
Comment (willclarktech) [inline test/api/liskApi.js]: `should recognise when a callback function is passed instead of an options object`

File: test/api/liskApi.js [hunk 37]
```diff
@@ -695,4 +694,45 @@ describe('Lisk.api()', () => {
...
+			address: '18160565574430594874',
+			limit: 0,
+			offset: 5,
+		};
+		const resolvedConflictObject = {
+			address: '18160565574430594874',
+			limit: requestDataLimit,
+			offset: requestOffset,
+		};
+
```
Comment (willclarktech) [inline test/api/liskApi.js]: Test when neither options nor callback is passed?

File: test/api/liskApi.js [hunk 38]
```diff
@@ -695,4 +694,45 @@ describe('Lisk.api()', () => {
...
+			const createObject = privateApi.constructRequestData({ address }, requestData);
+			(createObject).should.be.eql(expectedObject);
+		});
+
+		it('should construct with variable and callback', () => {
+			const createObject = privateApi.constructRequestData({ address }, () => true);
+			(createObject).should.be.eql({ address });
+		});
+
+		it('should give conflicting input, first parameters passed', () => {
```
Comment (willclarktech) [inline test/api/liskApi.js]: `should prioritise values from the data object when the data object and options object conflict`

File: test/api/liskApi.js (hunk 33 above)
Comment (Tosch110) [inline test/api/liskApi.js]: yes

File: test/api/liskApi.js [hunk 39]
```diff
@@ -5,12 +5,16 @@ describe('Lisk.api()', () => {
...
+	const defaultSecondSecret = 'second secret';
 	const GET = 'GET';
 	const POST = 'POST';
 	const defaultAddress = {
 		publicKey: '5d036a858ce89f844491762eb89e2bfbd50a4a0a0da658e4b2628b25b117ae09',
 		address: '18160565574430594874L',
 	};
+	const defaultRequestLimit = 10;
+	const defaultRequestOffset = 101;
+	const defaultAmount = 100000000;
```
Comment (willclarktech) [inline test/api/liskApi.js]: Let's make this easier to read with either a `fixedPoint` or just `e8`.

File: test/api/liskApi.js [hunk 40]
```diff
@@ -218,460 +222,177 @@ describe('Lisk.api()', () => {
...
-			LSK.listBlocks('1', callback);
+			it('should list standby delegates with a default offset and ordering when not specified', () => {
+				const options = {};
 
-			(LSK.sendRequest.calledWith('GET', 'blocks', options)).should.be.true();
-			(callback.called).should.be.true();
-			(callback.calledWith(expectedResponse)).should.be.true();
-			LSK.sendRequest.restore();
+				LSK.listStandbyDelegates(defaultRequestLimit, options, callback);
+				(LSK.sendRequest.calledWithExactly(GET, 'delegates', { limit: defaultRequestLimit, orderBy: 'rate:asc', offset: defaultRequestOffset }, callback)).should.be.true();
```
Comment (willclarktech) [inline test/api/liskApi.js]: `defaultOrderBy` instead of `'rate:asc'`.

File: test/api/liskApi.js [hunk 41]
```diff
@@ -218,460 +222,177 @@ describe('Lisk.api()', () => {
...
-		it('should get a block of certain height', () => {
-			const callback = sinon.spy();
-			const blockId = '2346638';
-			const options = { height: blockId };
-			sinon.stub(LSK, 'sendRequest').callsArgWith(3, expectedResponse);
 
-			LSK.getBlock(blockId, callback);
+		describe('#listBlocks', () => {
+			it('should list amount of blocks defined', () => {
+				const options = { limit: '1' };
```
Comment (willclarktech) [inline test/api/liskApi.js]: `defaultRequestLimit`

File: test/api/liskApi.js [hunk 42]
```diff
@@ -218,460 +222,177 @@ describe('Lisk.api()', () => {
...
-			(LSK.sendRequest.calledWith('GET', 'accounts/delegates', options)).should.be.true();
-			(callback.called).should.be.true();
-			(callback.calledWith(expectedResponse)).should.be.true();
-			LSK.sendRequest.restore();
+		describe('#listTransactions', () => {
+			it('should list transactions of a defined account', () => {
+				const recipientAddress = '12731041415715717263L';
+				const senderAddress = '15731041415715717263L';
+				const options = {
+					recipientId: recipientAddress,
```
Comment (willclarktech) [inline test/api/liskApi.js]: This should be removed from the `options` object passed to `listTransactions` (but not the object for assertion).

File: test/api/liskApi.js [hunk 43]
```diff
@@ -218,460 +222,177 @@ describe('Lisk.api()', () => {
...
-					generatorId: '6356913781456505636L',
-					blockSignature: '2156b5b20bd338fd1d575ddd8550fd5675e80eec70086c31e60e797e30efdeede8075f7ac35db3f0c45fed787d1ffd7368a28a2642ace7ae529eb538a0a90705',
-					confirmations: 1,
-					totalForged: '500000000',
-				}],
-			},
-		};
+		describe('#listStandbyDelegates', () => {
+			it('should list standby delegates', () => {
+				const options = { limit: defaultRequestLimit, orderBy: 'rate:asc', offset: defaultRequestOffset };
```
Comment (willclarktech) [inline test/api/liskApi.js]: This test should not use defaults.

File: test/api/liskApi.js [hunk 44]
```diff
@@ -218,460 +222,177 @@ describe('Lisk.api()', () => {
...
-			return liskApi().getMultisignatureTransaction('123', (result) => {
-				(result).should.be.ok();
-				(result).should.be.type('object');
+		describe('#getMultisignatureTransaction', () => {
+			it('should get a multisignature transaction by id', () => {
+				return liskApi().getMultisignatureTransaction('123', (result) => {
+					(result).should.be.ok();
+					(result).should.be.type('object');
+					(result).should.have.property('error').be.eql('Transaction not found');
+				});
```
Comment (willclarktech) [inline test/api/liskApi.js]: These tests should conform to the pattern of the other tests.

File: test/api/liskApi.js [hunk 45]
```diff
@@ -218,460 +222,177 @@ describe('Lisk.api()', () => {
...
-			(callback.called).should.be.true();
-			(callback.calledWith(expectedResponse)).should.be.true();
-			LSKnode.sendRequest.restore();
+				LSK.sendLSK(recipientId, defaultAmount, defaultSecret, defaultSecondSecret, callback);
+				(LSK.sendRequest.calledWithExactly(POST, 'transactions', {
+					recipientId,
+					amount: defaultAmount,
+					secret: defaultSecret,
+					secondSecret: defaultSecondSecret,
+				}, callback)).should.be.true();
```
Comment (willclarktech) [inline test/api/liskApi.js]: This object should be put in an `options` variable to conform to the pattern of the other tests.

File: test/api/liskApi.js [hunk 46]
```diff
@@ -695,4 +416,45 @@ describe('Lisk.api()', () => {
...
+			offset: 5,
+		};
+		const resolvedConflictObject = {
+			address: '18160565574430594874',
+			limit: defaultRequestLimit,
+			offset: defaultRequestOffset,
+		};
+
+		it('should merge a data object with an options object', () => {
+			const createObject = privateApi.constructRequestData({ address }, optionsObject);
```
Comment (willclarktech) [inline test/api/liskApi.js]: `const requestData` or `const constructedRequestData`

File: test/api/liskApi.js [hunk 47]
```diff
@@ -695,4 +416,45 @@ describe('Lisk.api()', () => {
...
+		const optionsObject = {
+			limit: defaultRequestLimit,
+			offset: defaultRequestOffset,
+		};
+		const expectedObject = {
+			address: '18160565574430594874L',
+			limit: defaultRequestLimit,
+			offset: defaultRequestOffset,
+		};
+		const conflictObject = {
```
Comment (willclarktech) [inline test/api/liskApi.js]: `const conflictingOptionsObject` or `optionsWithConflictObject`

File: test/api/liskApi.js [hunk 48]
```diff
@@ -695,4 +416,45 @@ describe('Lisk.api()', () => {
...
+
+		it('should recognise when a callback function is passed instead of an options object', () => {
+			const createObject = privateApi.constructRequestData({ address }, () => true);
+			(createObject).should.be.eql({ address });
+		});
+
+		it('should prioritise values from the data object when the data object and options object conflict', () => {
+			const createObject = privateApi.constructRequestData(
+				{ limit: defaultRequestLimit, offset: defaultRequestOffset }, conflictObject,
+			);
```
Comment (willclarktech) [inline test/api/liskApi.js]: It would make more sense to me to have something like this:
```js
privateApi.constructRequestData(
    { address: '123L', limit: 4 }, // data
    { address, limit: defaultRequestLimit, offset: defaultRequestOffset }, // options
);
```
I.e. we have some options with a normal address and some defaults, and we want to override it with a weirdo address and an unusual limit, but keep whatever default options we haven't explicitly overridden.

File: test/api/liskApi.js [hunk 49]
```diff
@@ -695,4 +416,45 @@ describe('Lisk.api()', () => {
 			});
 		});
 	});
+
+
+	describe('#constructRequestData', () => {
+		const address = defaultAddress.address;
```
Comment (willclarktech) [inline test/api/liskApi.js]: `const { address } = defaultAddress;`

File: test/api/liskApi.js [hunk 50]
```diff
@@ -218,460 +224,185 @@ describe('Lisk.api()', () => {
...
 
-			LSK.listActiveDelegates('1', callback);
-
-			(LSK.sendRequest.calledWith(GET, 'delegates/', options)).should.be.true();
-			(callback.called).should.be.true();
-			(callback.calledWith(expectedResponse)).should.be.true();
-			LSK.sendRequest.restore();
+		beforeEach(() => {
+			callback = sinon.spy();
+			sinon.stub(LSK, 'sendRequest');
```
Comment (willclarktech) [inline test/api/liskApi.js]: Prefer storing the stub in a variable for easier reference elsewhere. E.g.
```js
let stub;
beforeEach(() => {
    stub = sinon.stub(LSK, 'sendRequest');
});

afterEach(() => {
    stub.restore();
});
```

File: test/api/liskApi.js [hunk 51]
```diff
@@ -218,460 +224,185 @@ describe('Lisk.api()', () => {
...
 		});
-	});
 
-	describe('#getMultisignatureTransaction', () => {
-		it('should get a multisignature transaction by id', () => {
-			return liskApi().getMultisignatureTransaction('123', (result) => {
-				(result).should.be.ok();
-				(result).should.be.type('object');
+		describe('#listMultisignatureTransactions', () => {
+			it('should list all current not signed multisignature transactions', () => {
```
Comment (willclarktech) [inline test/api/liskApi.js]: `unsigned`

File: test/api/liskApi.js [hunk 52]
```diff
@@ -695,4 +426,45 @@ describe('Lisk.api()', () => {
...
+		});
+
+		it('should recognise when a callback function is passed instead of an options object', () => {
+			const requestData = privateApi.constructRequestData({ address }, () => true);
+			(requestData).should.be.eql({ address });
+		});
+
+		it('should prioritise values from the data object when the data object and options object conflict', () => {
+			const requestData = privateApi.constructRequestData(
+				{ limit: defaultRequestLimit, offset: defaultRequestOffset }, optionsWithConflictObject,
```
Comment (willclarktech) [inline test/api/liskApi.js]: This is the other way round to what I suggested, but it's ok with me since it tests the conflict.
